### PR TITLE
flatten measurements for satellite

### PR DESCRIPTION
**Pull request by ohnorobo**
Apologies for the size of this PR 😬  I promise >1000 of these lines are just tests. Probably the best way to understand what's happening here is to start by looking at the test cases to get a sense for what outputs we're aiming for.

This change breaks out individual DNS requests into their own rows. This works in different ways for v1, v2.1, and v2.2 data since they store different information about their measurements.

The other major internal change is how we handle received ip tagging. In the old code we split up rows per-received-ip to tag, then reconstructed (grouped) them by `measurement_id`. However that makes it impossible to separate out multiple rows per measurement, so now we're using an additional `roundtrip_id` per final output row. I've also changed some intermediate stages in the code where we were storing Dicts instead of Lists in the `received` field. Now even in intermediate stages we only store lists, which matches the final output schema.

Some changes I have not made:
- The final schema format for rcodes currently is List[str], ultimately we want it to just be Int (a single rcode) but I don't want to add on that schema change here.
- I'd also like to eventually update some of the field names in the schema to be clearer
- The received ip tagging currently works in tests, but fails when running full-size backfills. I haven't been able to get this fixed yet. Currently I plan to turn off the received ip tagging temporarily in a follow-up PR so that we're able to make faster progress on the dashboard.

### CODE REVIEW

File: pipeline/metadata/flatten_satellite.py [hunk 1]
```diff
@@ -75,111 +76,146 @@ def format_timestamp(timestamp: str) -> str:
...
 
 
-def _process_received_ips(
-    row: Row, received_ips: Optional[Dict[str, str]]) -> Iterator[Row]:
+def _process_received_ips_v1(
+    row: Row, received_ips: Optional[Dict[str, List[str]]]) -> Iterator[Row]:
   """Add received_ip metadata to a Satellite row
+
   Args:
     row: existing row of satellite data
```
Comment (fortuna) [inline pipeline/metadata/flatten_satellite.py:84]: Is `row` what we read from their file? What does it represent?

File: pipeline/metadata/flatten_satellite.py [hunk 2]
```diff
@@ -75,111 +76,146 @@ def format_timestamp(timestamp: str) -> str:
...
-              [tag for tag in matched if tag in SATELLITE_TAGS])
-        row['received'].append(received)
-      yielded = True
-      yield row.copy()
-    # Do not output rows if there are no trials for the test domain.
-  if not yielded and has_test_domain:
-    # All trials for the test domain in `scan` are errors
-    yield row.copy()
+  row['controls_failed'] = not scan['passed_control']
+  received_ips = scan.get('response').copy()
```
Comment (fortuna) [inline pipeline/metadata/flatten_satellite.py:165]: Rename `received_ips` to `response`.

File: pipeline/metadata/flatten_satellite.py [hunk 3]
```diff
@@ -75,111 +76,146 @@ def format_timestamp(timestamp: str) -> str:
...
+
+  if not received_ips:
+    row['received'] = []
+    yield row
+    return
+
+  # separate into one answer ip per row for tagging
+  rcodes = received_ips.pop('rcode', [])
```
Comment (fortuna) [inline pipeline/metadata/flatten_satellite.py:173]: Why would you have more than one rcode and error in a response?

This is very hard to follow. I propose a logic change to first:
- extract all responses from the Satellite data. Each response must only have one rcode and error.
- Annotate and output the responses.

The first step changes with the data version, but the second should be shared across versions.

Introduce functions like get_responses_v1 for example. That may even be its own step in the pipeline.

File: pipeline/metadata/flatten_satellite.py [hunk 4]
```diff
@@ -371,54 +463,60 @@ def _process_satellite_v2_control(
...
-              format_timestamp(responses[-1]['end_time']),
+              format_timestamp(response['end_time']),
           'anomaly':
               None,
           'success':
               not scan['connect_error'],
           'controls_failed':
               not scan['passed_control'],
-          'rcode': [str(response['rcode']) for response in responses],
+          'rcode': [str(response['rcode'])],
```
Comment (fortuna) [inline pipeline/metadata/flatten_satellite.py:494]: This should not be a list.

File: pipeline/metadata/flatten_satellite.py [hunk 5]
```diff
@@ -371,54 +463,60 @@ def _process_satellite_v2_control(
...
-      # always contains some non-control trial domains
-      is_control_domain = False
+
+    for response in responses:
+      is_control_domain = flatten_base.is_control_url(response['url'])
 
       row = {
           'domain':
-              scan['test_url'],
+              response['url'],
```
Comment (fortuna) [inline pipeline/metadata/flatten_satellite.py:472]: Extract the domain from the URL.
Why is the field called URL?

File: pipeline/metadata/flatten_satellite.py (hunk 3 above)
Comment (fortuna) [reply pipeline/metadata/flatten_satellite.py:173]: The flow should be:
input entry -> responses -> annotated responses
Having the data for multiple responses in one merged output row is not analyzable in a sane way. We can't do that.

**Review by fortuna — CHANGES_REQUESTED** [review]

File: pipeline/metadata/flatten_satellite.py [hunk 6]
```diff
@@ -75,111 +76,146 @@ def format_timestamp(timestamp: str) -> str:
...
+    received_ips: data on the response from the resolver
+      ex: {<ip>: ["tags"]}
+
   Yields:
-    Rows
+    Row with an additional 'received' array field
   """
-  if not received_ips:
+  if received_ips is None:
+    row['received'] = []
```
Comment (fortuna) [inline pipeline/metadata/flatten_satellite.py:92]: Rename `row` to `base_response`.

File: pipeline/metadata/flatten_satellite.py (hunk 2 above)
Comment (fortuna) [reply pipeline/metadata/flatten_satellite.py:165]: Actually, maybe we should rename `row` to `base_observation` and `received_ips` to something like `input_response_data`, since it's not a response, and it's not just IPs.

I'm using "observation", instead of "responses" because sometimes the observation is not a response (e.g. timeout)

File: pipeline/metadata/flatten_satellite.py [hunk 7]
```diff
@@ -75,111 +76,146 @@ def format_timestamp(timestamp: str) -> str:
...
+
+  if successful_test_rcode:
+    all_received = []
+    for (ip, tags) in received_ips.items():
+      received = {'ip': ip}
+      if tags:
+        received['matches_control'] = _append_tags(tags)
+      all_received.append(received)
+
+    success_row = row.copy()
```
Comment (fortuna) [inline pipeline/metadata/flatten_satellite.py:199]: Rename `success_row` to `observation`. It should represent a single DNS request + observation.

File: pipeline/metadata/flatten_satellite.py [hunk 8]
```diff
@@ -75,111 +76,146 @@ def format_timestamp(timestamp: str) -> str:
...
+
+  if any(rcode == "0" for rcode in failed_test_rcodes):
+    raise Exception(
+        f"Satellite v2.1 measurement has multiple 0 rcodes: {filepath} - {scan}"
+    )
+
+  if not successful_test_rcode and received_ips:
+    # TODO figure out how to handle this bug case from 2021-05-16
+    # for now drop the measurements
+    logging.warning(
```
Comment (fortuna) [inline pipeline/metadata/flatten_satellite.py:186]: Perhaps introduce an error counter

File: pipeline/metadata/flatten_satellite.py [hunk 9]
```diff
@@ -75,111 +76,146 @@ def format_timestamp(timestamp: str) -> str:
...
+        f"Satellite v2.1 measurement has ips but no 0 rcode: {filepath} - {scan}"
+    )
+    return
+
+  if successful_test_rcode:
+    all_received = []
+    for (ip, tags) in received_ips.items():
```
Comment (fortuna) [inline pipeline/metadata/flatten_satellite.py:193]: Isn't received_ips a dict with fields like "rcode"?

File: pipeline/metadata/flatten_satellite.py [hunk 10]
```diff
@@ -75,111 +76,146 @@ def format_timestamp(timestamp: str) -> str:
...
+        received['matches_control'] = _append_tags(tags)
+      all_received.append(received)
+
+    success_row = row.copy()
+    success_row['received'] = all_received
+    success_row['rcode'] = successful_test_rcode
+    yield success_row
+
+  for rcode in failed_test_rcodes:
+    failed_row = row.copy()
```
Comment (fortuna) [inline pipeline/metadata/flatten_satellite.py:205]: I'd call this `observation`

File: pipeline/metadata/flatten_satellite.py [hunk 11]
```diff
@@ -75,111 +76,146 @@ def format_timestamp(timestamp: str) -> str:
...
+    failed_row['rcode'] = [rcode]
+    # -1 rcodes corrospond to the error messages in order.
+    if rcode == "-1":
+      failed_row['error'] = errors.pop() if errors else None
+    yield failed_row
+
+  # There is a bug in some v2.1 data where -1 rcodes weren't recorded.
+  # In that case we add them back in manually
+  for error in errors:
+    error_row = row.copy()
```
Comment (fortuna) [inline pipeline/metadata/flatten_satellite.py:215]: Rename to `observation`. In this case we can't really call `response`, because there was no dns response. (though perhaps you got an ICMP error)

File: pipeline/metadata/flatten_satellite.py [hunk 12]
```diff
@@ -75,111 +76,146 @@ def format_timestamp(timestamp: str) -> str:
...
+  if successful_test_rcode:
+    all_received = []
+    for (ip, tags) in received_ips.items():
+      received = {'ip': ip}
+      if tags:
+        received['matches_control'] = _append_tags(tags)
+      all_received.append(received)
+
+    success_row = row.copy()
+    success_row['received'] = all_received
```
Comment (fortuna) [inline pipeline/metadata/flatten_satellite.py:200]: Rename `received` to `answer`. The rcode you also "received".

File: pipeline/metadata/flatten_satellite.py [hunk 13]
```diff
@@ -251,7 +287,7 @@ def _process_satellite_v1(self, date: str, scan: Any, filepath: str,
         'error': scan.get('error', None),
         'anomaly': not scan['passed'] if 'passed' in scan else None,
         'success': 'error' not in scan,
-        'received': None,
+        'received': [],
```
Comment (fortuna) [inline pipeline/metadata/flatten_satellite.py:290]: `answer`

**Review by fortuna — CHANGES_REQUESTED** [review]

File: pipeline/metadata/flatten_satellite.py (hunk 4 above)
Comment (ohnorobo) [reply pipeline/metadata/flatten_satellite.py:494]: That will be changed in https://github.com/censoredplanet/censoredplanet-analysis/pull/117, but not this PR.

File: pipeline/metadata/flatten_satellite.py (hunk 13 above)
Comment (ohnorobo) [reply pipeline/metadata/flatten_satellite.py:290]: schema name changes will happen in https://github.com/censoredplanet/censoredplanet-analysis/pull/119

File: pipeline/metadata/flatten_satellite.py (hunk 12 above)
Comment (ohnorobo) [reply pipeline/metadata/flatten_satellite.py:200]: https://github.com/censoredplanet/censoredplanet-analysis/pull/119

File: pipeline/metadata/flatten_satellite.py [hunk 14]
```diff
@@ -75,111 +76,146 @@ def format_timestamp(timestamp: str) -> str:
...
+
   Args:
     row: existing row of satellite data
-    received_ips: data on the response from the resolver (error or ips)
+    received_ips: data on the response from the resolver
+      ex: {<ip>: ["tags"]}
+
   Yields:
-    Rows
+    Row with an additional 'received' array field
```
[the remark this answers is not in the document]
Comment (ohnorobo) [reply pipeline/metadata/flatten_satellite.py:89]: I agree that would be very helpful, but that's a big refactor that can't happen in this PR.

File: pipeline/metadata/flatten_satellite.py (hunk 11 above)
Comment (ohnorobo) [reply pipeline/metadata/flatten_satellite.py:215]: Changed to `error_observation`

File: pipeline/metadata/flatten_satellite.py (hunk 10 above)
Comment (ohnorobo) [reply pipeline/metadata/flatten_satellite.py:205]: Changed to `failed_observation`

File: pipeline/metadata/flatten_satellite.py (hunk 9 above)
Comment (ohnorobo) [reply pipeline/metadata/flatten_satellite.py:193]: In v2.1 the response field looks like this:
3hksuhqzaqs2h7k and contains results from several roundtrip tests.

https://censoredplanet.slack.com/archives/C01DK5FF4KD/p1639594115136600

So we need to untangle all those fields into rows where each represents an individual roundtrip. First we pop off the rcode/error fields, then we treat the remainder as a dict {'ip': ['tag_a', 'tag_b']}

File: pipeline/metadata/flatten_satellite.py (hunk 8 above)
Comment (ohnorobo) [reply pipeline/metadata/flatten_satellite.py:186]: We can count these logs in a job that succeeded. There are 4070 rows that trigger this error in the whole dataset.

6fGL4r5iN8nsVjf

https://console.cloud.google.com/logs/query;query=resource.type%3D%22dataflow_step%22%0Aresource.labels.job_id%3D%222022-01-17_20_58_59-987126952126004496%22%0Aseverity%3DWARNING%0A%22Satellite%20v2.1%20measurement%20has%20ips%20but%20no%200%20rcode%22;timeRange=2022-01-18T04:58:59.933Z%2F2022-01-18T13:01:53.036Z;cursorTimestamp=2022-01-18T11:06:33.445386171Z?project=firehook-censoredplanet

We don't expect this number to ever get any larger (since no more 2.1 data is being generated.) So this number will never get bigger. Implementing a counter in code doesn't seem like a good use of time.

File: pipeline/metadata/flatten_satellite.py (hunk 7 above)
Comment (ohnorobo) [reply pipeline/metadata/flatten_satellite.py:199]: renamed to `success_observation`. That is what it represents.

Comment (ohnorobo) [reply pipeline/metadata/flatten_satellite.py:199]: renamed `row` to `base_observation` and `received_ips` to `input_response_data`

File: pipeline/metadata/flatten_satellite.py (hunk 6 above)
Comment (ohnorobo) [reply pipeline/metadata/flatten_satellite.py:92]: done

File: pipeline/metadata/flatten_satellite.py (hunk 2 above)
Comment (ohnorobo) [reply pipeline/metadata/flatten_satellite.py:165]: renamed `row` to `base_observation` and `received_ips` to `input_response_data`

File: pipeline/metadata/flatten_satellite.py [hunk 15]
```diff
@@ -75,111 +76,146 @@ def format_timestamp(timestamp: str) -> str:
   return f'{date}T{time}{timezone_hour}:{timezone_seconds}'
 
 
-def _process_received_ips(
-    row: Row, received_ips: Optional[Dict[str, str]]) -> Iterator[Row]:
+def _process_received_ips_v1(
```
[the remark this answers is not in the document]
Comment (ohnorobo) [reply pipeline/metadata/flatten_satellite.py:79]: renamed to `_annotate_received_ips_v1`

File: pipeline/metadata/beam_metadata.py [hunk 16]
```diff
@@ -53,10 +53,12 @@ def merge_metadata_with_rows(  # pylint: disable=unused-argument
     new_row.update(row)
     if field == 'received':
       if new_row['received']:
-        new_row['received'].update(ip_metadata)
-        new_row['received'].pop('date', None)
-        new_row['received'].pop('name', None)
-        new_row['received'].pop('country', None)
+        # Double-flattened rows are stored with a single received ip in each list
+        # to be reconstructed later
+        new_row['received'][0].update(ip_metadata)
```
[the remark this answers is not in the document]
Comment (ohnorobo) [reply pipeline/metadata/beam_metadata.py:58]: The way we're doing the received ip tagging is:
-  there's a flattened (representing a single observation) row with multiple received ips. 
- We then "double flatten" that into several rows each with a single received ip each (and the same `roundtrip_id`)
- we then add tags and do the co-group with metadata. (the co-group is why this whole dance is required)
- Then we reconstruct the original measurement with multiple received_ips [like this](https://github.com/censoredplanet/censoredplanet-analysis/blob/fb0d8a5b4dfa90edeca2347c7141f0d5f331c085/pipeline/metadata/satellite.py#L376).

We may want to make some changes here, but I didn't want to do that while also doing the flattening. If the row has made it to this point in the code its `received` field will be a list with a single ip like [{'ip': '1.2.3.4'}] ready to have metadata added.

File: pipeline/metadata/flatten_satellite.py (hunk 1 above)
Comment (ohnorobo) [reply pipeline/metadata/flatten_satellite.py:84]: `row` is the python object we are constructing in order to write it as a row in bigquery. Data read from the scan file is usually called `line` (if it's a json string) or `scan` if we've parsed the json into a python object.

File: pipeline/metadata/flatten_satellite.py (hunk 3 above)
Comment (ohnorobo) [reply pipeline/metadata/flatten_satellite.py:173]: I'm not merging multiple responses into one row here. I'm trying to tease apart multiple responses that are already merged in the scan data into individual rows which each represent a single observation.

I do agree that this code is hard to understand though. Maybe we should get on a call and do a code walkthrough to see where there's bad flow/lack of needed comments explaining the data format?

File: pipeline/metadata/flatten_satellite.py (hunk 14 above)
[the remark this answers is not in the document]
Comment (ohnorobo) [reply pipeline/metadata/flatten_satellite.py:90]: The options seem to be using either NamedTuple or the new [dataclass](https://www.python.org/dev/peps/pep-0557/)

File: pipeline/metadata/flatten_satellite.py [hunk 17]
```diff
@@ -371,54 +464,60 @@ def _process_satellite_v2_control(
         Rows
     """
     responses = scan.get('response', [])
```
Comment (fortuna) [inline pipeline/metadata/flatten_satellite.py:466]: Please rename `scan` to `responses_entry`. Introduce a `ResponsesEntry` type that corresponds to the responses.json and responses_control.json files (as defined at https://docs.censoredplanet.org/dns.html)

File: pipeline/metadata/flatten_satellite.py [hunk 18]
```diff
@@ -371,54 +464,60 @@ def _process_satellite_v2_control(
...
           'start_time':
-              format_timestamp(responses[0]['start_time']),
+              format_timestamp(response['start_time']),
           'end_time':
-              format_timestamp(responses[-1]['end_time']),
+              format_timestamp(response['end_time']),
           'anomaly':
               None,
           'success':
               not scan['connect_error'],
```
Comment (fortuna) [inline pipeline/metadata/flatten_satellite.py:492]: I don't see `connect_error` and 'passed_control` in the `responses.json` type definition.

File: pipeline/metadata/flatten_satellite.py (hunk 17 above)
Comment (fortuna) [reply pipeline/metadata/flatten_satellite.py:466]: Also specify in the comment the filename this is coming from.

File: pipeline/metadata/flatten_satellite.py [hunk 19]
```diff
@@ -371,54 +464,60 @@ def _process_satellite_v2_control(
...
-              row['received'] = {'ip': ip}
-              yield row.copy()
+
+      if response['error'] == 'null':
+        error = None
+      else:
+        error = response['error']
+      row['error'] = error
+
+      received_ips = response['response']
```
Comment (fortuna) [inline pipeline/metadata/flatten_satellite.py:510]: I don't see `response` in `ResponsesEntry.response`.  Did you mean `answer`?

**Review by fortuna — CHANGES_REQUESTED** [review]

File: pipeline/metadata/flatten_satellite.py (hunk 18 above)
Comment (ohnorobo) [reply pipeline/metadata/flatten_satellite.py:492]: This code is heavily unit tested (with the unit tests coming from example measurements) and has been run against the entire satellite dataset. If you're seeing fields that don't match the documentation it's probably that the docs are wrong.

Here's the example measurement in the unit tests with this field https://github.com/censoredplanet/censoredplanet-analysis/blob/16276c030dc9482c3c0601850f76503001c36cb4/pipeline/metadata/test_satellite.py#L560

File: pipeline/metadata/flatten_satellite.py (hunk 5 above)
Comment (ohnorobo) [reply pipeline/metadata/flatten_satellite.py:473]: I'm happy to do either of those name changes. But both of them require a schema change, so they have to happen in https://github.com/censoredplanet/censoredplanet-analysis/pull/119.

Currently, I need to preserve the full content of this field so I'll be able to investigate it and figure out why these are appearing as urls at all. If I do the extraction here (in this PR) then that data will become impossible to access.

Comment (fortuna) [reply pipeline/metadata/flatten_satellite.py:473]: I don't see how adding a column would be a problem. It actually concerns me that we may not be aggregating correctly if it's urls and not domains.
I'm fine with punting on that for now, but please add a TODO

File: pipeline/metadata/flatten_satellite.py (hunk 19 above)
Comment (ohnorobo) [reply pipeline/metadata/flatten_satellite.py:510]: Same as above. Here are the example test with the `response.response` field

https://github.com/censoredplanet/censoredplanet-analysis/blob/534652d4ea0209d49a176e1c9ddbee5bfeae7a67/pipeline/metadata/test_flatten_satellite.py#L189

File: pipeline/metadata/flatten_satellite.py (hunk 17 above)
Comment (ohnorobo) [reply pipeline/metadata/flatten_satellite.py]: done and done

File: pipeline/metadata/satellite.py [hunk 20]
```diff
@@ -396,14 +396,52 @@ def _unflatten_satellite(flattened_measurement: Iterable[Row]) -> Iterator[Row]:
...
     # during the tagging step if a vantage point also appears as a response IP.
     for tag in ['asname', 'asnum', 'http', 'cert']:
       combined.pop(tag, None)
+
+    combined.pop('roundtrip_id')  # Remove roundtrip id
     yield combined
 
 
+def flatten_received_ips(row: Row) -> Iterator[Row]:
+  """Flatten a row with multiple received ips into rows with a single ip.
```
[the remark this answers is not in the document]
Comment (ohnorobo) [reply pipeline/metadata/satellite.py:409]: Changed the name to dns_roundtrip

changed the <dict> abstraction in the comments to be {'ip': "1.2.3.4'} which matches the structure of the actual data.

File: pipeline/metadata/flatten_satellite.py (hunk 5 above)
Comment (ohnorobo) [reply pipeline/metadata/flatten_satellite.py:492]: TODO added

File: pipeline/metadata/satellite.py [hunk 21]
```diff
@@ -396,14 +396,60 @@ def _unflatten_satellite(flattened_measurement: Iterable[Row]) -> Iterator[Row]:
...
+  dns_roundtrip['roundtrip_id'] = uuid.uuid4().hex
+
+  all_received = dns_roundtrip['received']
+
+  if len(all_received) == 0:
+    yield dns_roundtrip
+    return
+
+  for received in all_received:
+    dns_roundtrip['received'] = [received]
```
Comment (fortuna) [inline pipeline/metadata/satellite.py:449]: This is type abuse. The new field is no longer a list, and the structure has a different semantics. You need to define a new type. But see my refactoring suggestion above. By not carrying the entire observation around, you are forced to not reuse the type.

File: pipeline/metadata/satellite.py [hunk 22]
```diff
@@ -396,14 +396,60 @@ def _unflatten_satellite(flattened_measurement: Iterable[Row]) -> Iterator[Row]:
...
+            'ip': '1.2.3.4'
+        },{
+            'ip': '5.6.7.8'
+        }]
+      }
+
+  Returns:
+    Rows like
+      {
+        "field": "value"
```
[the remark this answers is not in the document]
Comment (fortuna) [reply pipeline/metadata/satellite.py:425]: BTW, the transform doesn't have to output two collections. You can run two separate transforms on the same input instead, which is cleaner.

So one transform would do:
dsn_roundtrips -> (roundtrip_id, dns_roundtrip)

and the other would do:
dns_roundtrips -> (roundtrip_id, answer_ip)

File: pipeline/metadata/satellite.py [hunk 23]
```diff
@@ -421,13 +467,19 @@ def process_satellite_with_tags(
   rows = (
       row_lines | 'flatten json' >> beam.ParDo(
           flatten.FlattenMeasurement()).with_output_types(Row))
+
+  # PCollection[Row] each with only a single element in the received arrays
+  received_ip_flattened_rows = (
+      rows | 'flatten received ips' >>
```
Comment (fortuna) [inline pipeline/metadata/satellite.py:473]: This is where you can create the two separate PCollections by running two different transforms on the `rows`.

**Review by fortuna — CHANGES_REQUESTED** [review]

**Review by fortuna — APPROVED** [review]
Let's add a TODO for the triple CoGroup

**Comment by ohnorobo** [comment]
Added a TODO about the triple cogroup in [8de0daa](https://github.com/censoredplanet/censoredplanet-analysis/pull/116/commits/8de0daaffeee1f4da56a1c33a2698222327bd03b)

As discussed offline I'm going to merge this now and work on some incremental refactoring in future PRs.
